### PR TITLE
Chalkify - Tweak `Clause` definition and HRTBs

### PR DESCRIPTION
**Pull request by scalexm**
r? @nikomatsakis

### CODE REVIEW

**Review by tmandry — COMMENTED** [review]
`PolyDomainGoal` definitely seems like a good approach from my experience with #49435. I'll try rebasing on your branch.

File: src/librustc_traits/lowering.rs [hunk 1]
```diff
@@ -134,13 +125,16 @@ fn program_clauses_for_trait<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefI
...
     let impl_trait = DomainGoal::Holds(WhereClauseAtom::Implemented(trait_pred));
 
     // `Implemented(Self: Trait<P1..Pn>) :- FromEnv(Self: Trait<P1..Pn>)`
-    let clause = Clause::Implies(vec![from_env], impl_trait);
-    Lrc::new(vec![clause])
+    let clause = ProgramClause {
+        goal: impl_trait,
+        hypotheses: vec![from_env],
+    };
+    Lrc::new(vec![Clause::ForAll(ty::Binder::dummy(clause))])
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc_traits/lowering.rs]: In chalk, clauses which come from the program are always fully quantified so as to account for generic parameters, e.g.:
```rust
impl<T> Foo for T {}
```
would lower to the following clause:
```
forall<T> { T: Foo }
```
In rustc, there is no concept of « quantified by type », only by lifetime (and quantification by lifetime is only used for higher ranked types like in `for<‘a> fn(&’a i32)`).

So the binder in front of the clause is indeed not useful *for the moment*, but you can view it as just a placeholder until rustc binders are reworked in order to account for types.

File: src/librustc_traits/lowering.rs [hunk 2]
```diff
@@ -184,14 +181,19 @@ struct ClauseDumper<'a, 'tcx: 'a> {
...
         for attr in attrs {
             if attr.check_name("rustc_dump_program_clauses") {
                 let clauses = self.tcx.program_clauses_for(def_id);
                 for clause in &*clauses {
-                    self.tcx.sess.struct_span_err(attr.span, &format!("{}", clause)).emit();
+                    let program_clause = match clause {
+                        Clause::Implies(program_clause) => program_clause,
+                        Clause::ForAll(program_clause) => program_clause.skip_binder(),
+                    };
+                    // Skip the top-level binder for a less verbose output
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc_traits/lowering.rs]: Yeah definitely, I’ll fix that :)

File: src/librustc/traits/mod.rs [hunk 3]
```diff
@@ -294,20 +296,39 @@ impl<'tcx> From<DomainGoal<'tcx>> for Goal<'tcx> {
...
+    Implies(ProgramClause<'tcx>),
+    ForAll(ty::Binder<ProgramClause<'tcx>>),
+}
+
+/// A "program clause" has the form `D :- G1, ..., Gn`. It is saying
+/// that the domain goal `D` is true if `G1...Gn` are provable. This
+/// is equivalent to the implication `G1..Gn => D`; we usually write
+/// it with the reverse implication operator `:-` to emphasize the way
+/// that programs are actually solved (via backchaining, which starts
+/// with the goal to solve and proceeds from there).
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc/traits/mod.rs]: (Actually @nikomatsakis’s comment from his chalkify-engine branch :p)

**Review by nikomatsakis — APPROVED** [review]
Did you want to fix [tmandry's nit](https://github.com/rust-lang/rust/pull/49497#discussion_r178406876)?

**Comment by scalexm** [comment]
@nikomatsakis done ✅

**Pull request merged**